### ISSUE TITLE
[Unitary Hack]Mix state evolution, and expectation value calculation

### ISSUE THREAD

**Issue by hongyehu** [hongyehu/PyClifford#22]
Implement mixed state quantum evolution, measurement, and expectation values. In the current implementation, the stabilizer state class has an attribute `stabilizer_state.r` which labels how many stabilizer generators are used to describe the state. For example, for a `N` qubit stabilizer state, in the `gs` matrix, the first `N` rows denote the stabilizer generators in binary vectors, and the second `N` rows denote the destabilizer generators in binary vectors. And the stabilizer generators used are from `stabilizer_state.r` to `N`, while destabilizer generators used are from `stabilizer_state.r+N` to `2N`. 

Therefore, if `stabilizer_state.r=0`, the state is fully stabilized by the first `N` generators, and the state is pure. And one need to change the evolution, measurement, calculation of expectation functions accordingly to general `0<stabilizer_state.r<=N` for the mixed states.

**Comment by vietnguyen012** [comment]
Hi, I would like to take up the issue. Can you assign it to me? thank you.

**Comment by vietnguyen012** [comment]
> Hi, I would like to take up the issue. Can you assign it to me? thank you.

P.S: I've just read the underlying code of pyclifford for the pure state operations, I understand nothing tbh 🙁. Do you have documents or papers explaining the algorithms for doing the projection's evolution, measurement and expectation for pure state or even better mixed states? I really appreciate your help! I am still new to the stabilizer code but I really want to learn more about it.

**Comment by hongyehu** [comment]
@vietnguyen012 according to the unitary fund, the bounty is not assigned to a person until one finishes it. I am happy to help you understand better. I will attach some references on the stabilizer state and its simulation here. If you need extra help, please let me know. :-)

**Comment by vietnguyen012** [comment]
yes, I will try to solve it. I am still waiting for your references. Thank you!

**Comment by vietnguyen012** [comment]
Hi, I noticed that it looks like the measurement function has already implemented measurements with r > 0. Is that right? I need to confirm this.

**Comment by dobbse42** [comment]
Hello! I'll be giving this one a shot. In your problem statement you mention that the stabilizer generators used go from StabilizerState.r to N, and the destabilizer generators from StabilizerState.r+N to 2N. This differs from the original Aaronson and Gottesman paper in two significant ways: 1) Aaronson and Gottesman place the destabilizers first, and the stabilizers second (I don't think this causes any issues, it's just important to remember when trying to compare the code to the algorithm outlined in the paper). 2) In the mixed case, the stabilizer mixed state is generated by r stabilizers and additionally has r destabilizers, not n-r as you have in the code. There are instead n-r logical X and logical Z operators (see section VIIA on page 11). I believe this to be a bug in your code which has simply not presented itself up to now since r has always been 0. I'm going to start working on implementing the modified operations with the assumption that gs will contain r stabilizers, followed by n-r X operators, followed by r destabilizers, followed by n-r Z operators, but actually making that change might affect other parts of the code. Have I misunderstood anything here? Thanks in advance.

**Comment by dobbse42** [comment]
After spending a while looking at the code, I am now under the impression that all of the requested functionality has actually already been implemented since the first commit. `stabilizer_project`, `stabilizer_measure`, and `stabilizer_expect` all allow for non-zero r and (I believe) function correctly. The only functionality which I believe may remain to be implemented is actually finding the logical Z and X operators (the "standby operators") when creating a stabilizer mixed state, as I cannot tell if the stabilizer_project function would halt at non-zero r with correct operator values if given stabilizers defining a stabilizer mixed state (and I don't think any of the existing unit tests look for this).

Is there still some desired functionality missing from the stabilizer functions in utils.py? Am I correct in understanding that the existing unit tests do not test for correct values of standby operators for arbitrary stabilizer mixed states, and so this is all that needs to be implemented?

Additionally, in the interest of helping others who might want to contribute to this project: `stabilizer_state.r` refers to the value "n-r" in the Aaronson and Gottesman paper (the number of logical operators, not the number of stabilizer generators), and `stabilizer_state.gs` holds the Z operators, followed by the stabilizer generators, followed by the X operators, followed by the destabilizer generators.

**Comment by vietnguyen012** [comment]
Hi @dobbse42, do you think the r need to be 0 after the algorithm is finished bc we basically purify the the mixed states by representing it as a tableau of a pure state? on the side note, I find it hard to reconcile the algorithm from the paper and pyclifford

**Comment by QChemEC** [comment]
Is this still open? I would like to work on that
